pkg/scheduler/core/generic_scheduler.go

```diff
@@ -483,6 +483,10 @@ func podFitsOnNode(
 					// we need to execute predicate functions since equivalence cache does not work
 					fit, reasons, err = predicate(pod, metaToUse, nodeInfoToUse)
 					if err != nil {
+						// TODO(harry): We need to unlock ecache before return, but find a better way to do this.
```

Comment:
@resouer I think we should create an inline function whose body covers lines 474 to 508 (and I think the function should return `fit` and `reasons`). Then we should remove all these `ecache.Unlock()` calls and add `defer ecache.Unlock()` after each `ecache.Lock()`.

Reply:
Thanks! This is exactly what I am doing. While I added a explicit function named `doPredicate`, not sure what do you think it VS anonymous one?